File: sdk-ios/Sources/WailoSDK/WailoScriptStore.swift
import Foundation
import WailoProtocol

/// Match metadata for daemon-owned Scripts. Source and execution never enter the host app.
final class WailoScriptStore: @unchecked Sendable {

    static let shared = WailoScriptStore()

    private let lock = NSLock()
    private var rules: [ScriptRule] = []
    private var snapshotOwner: WailoSnapshotOwner?
    private var snapshotGeneration: UInt64 = 0

    func replace(_ rules: [ScriptRule], owner: WailoSnapshotOwner? = nil) {
        lock.lock()
        defer { lock.unlock() }
        if let owner {
            guard owner.generation >= snapshotGeneration else { return }
            snapshotGeneration = owner.generation
        } else {
            snapshotGeneration = 0
        }
        snapshotOwner = owner
        self.rules = rules
    }

    func reset(owner: WailoSnapshotOwner? = nil) {
        lock.lock()
        defer { lock.unlock() }
        if let owner {
            guard snapshotOwner === owner else { return }
        } else {
            snapshotGeneration = 0
        }
        snapshotOwner = nil
        rules = []
    }

    func matches(url: String, method: String, phase: ScriptPhase) -> Bool {
        lock.lock()
        let snapshot = rules
        lock.unlock()
        return snapshot.contains { rule in
            rule.enabled
                && ((phase == .SCRIPT_PHASE_REQUEST && rule.on_request)
                    || (phase == .SCRIPT_PHASE_RESPONSE && rule.on_response))
                && (rule.methods.isEmpty
                    || rule.methods.contains { $0.caseInsensitiveCompare(method) == .orderedSame })
                && wildcardMatches(rule.url_pattern, url)
        }
    }

    private func wildcardMatches(_ pattern: String, _ text: String) -> Bool {
        let escaped = NSRegularExpression.escapedPattern(for: pattern)
        let body = escaped.replacingOccurrences(of: "\\*", with: ".*")
        return text.range(of: "^" + body + "$", options: [.regularExpression]) != nil
    }
}

File: sdk-android/src/main/kotlin/com/venbiasa/wailo/sdk/android/WailoScriptStore.kt
package com.venbiasa.wailo.sdk.android

import com.venbiasa.wailo.protocol.ScriptPhase
import com.venbiasa.wailo.protocol.ScriptRule

/** Match metadata for daemon-owned Scripts. Source and execution never enter the host app. */
object WailoScriptStore {
    private data class Snapshot(
        val owner: Any?,
        val rules: List<ScriptRule>,
    )

    private val lock = Any()

    @Volatile
    private var snapshot = Snapshot(owner = null, rules = emptyList())

    fun replace(rules: List<ScriptRule>) {
        synchronized(lock) {
            snapshot = Snapshot(owner = null, rules = rules)
        }
    }

    internal fun replace(rules: List<ScriptRule>, owner: Any) {
        synchronized(lock) {
            snapshot = Snapshot(owner = owner, rules = rules)
        }
    }

    internal fun reset(owner: Any) {
        synchronized(lock) {
            if (snapshot.owner === owner) snapshot = Snapshot(owner = null, rules = emptyList())
        }
    }

    fun matches(url: String, method: String, phase: ScriptPhase): Boolean =
        snapshot.rules.any { rule ->
            rule.enabled &&
                when (phase) {
                    ScriptPhase.SCRIPT_PHASE_REQUEST -> rule.on_request
                    ScriptPhase.SCRIPT_PHASE_RESPONSE -> rule.on_response
                    else -> false
                } &&
                methodMatches(rule.methods, method) &&
                urlWildcardMatches(rule.url_pattern, url)
        }
}
